src/numeric_util.rs

```diff
@@ -48,11 +48,11 @@ where
 
     // make it clear to the optimizer that this loop is short
     // and can not be autovectorized.
-    for i in 0..xs.len() {
+    for (i, x) in xs.iter().enumerate() {
```

Comment:
I assume this is fine, but we need to check that it still compiles to clean assembly.

Reply:
Great - how do you generally do this? I tried running `cargo asm ndarray::numeric_util::unrolled_fold --rust` but it couldn't find the function. (I can look more but figured you might know off-hand)

Reply:
I like using Matt Godbolt's Compiler Explorer (https://godbolt.org/) for standalone functions like this. (Make sure to switch the language to Rust.) Note that in order to get any output, you need to have a public, non-generic function. So, you could use this as the input:

```rust
pub fn foo(xs: &[f32]) -> f32 {
    unrolled_fold(xs, || 0., std::ops::Add::add)
}

pub fn unrolled_fold<A, I, F>(mut xs: &[A], init: I, f: F) -> A
where
    A: Clone,
    I: Fn() -> A,
    F: Fn(A, A) -> A,
{
    ...
}
```

to get the assembly for `foo`. You can specify the compiler options in the upper part of the right pane. It's worth trying `-O` or, for maximum optimization, `-C opt-level=3 -C target-cpu=native`.

I'm not very familiar with `cargo asm`, but I tried the command you suggested and got an error too. The error message for `cargo asm` says that the function must be present in the final binary. An easy way to ensure this is to add a function that is not generic and is part of the public API. So, for example, you could add this to `lib.rs`:

```rust
pub fn unrolled_sum_f32(xs: &[f32]) -> f32 {
    numeric_util::unrolled_fold(xs, num_traits::Zero::zero, std::ops::Add::add)
}
```

and then use `cargo asm ndarray::unrolled_sum_f32 --rust`.

The results I got in Compiler Explorer showed that the assembly for

```rust
    for (i, x) in xs.iter().enumerate() {
        if i >= 7 {
            break;
        }
        acc = f(acc.clone(), x.clone())
    }
```

was comparable to that for

```rust
    for i in 0..xs.len() {
        if i >= 7 {
            break;
        }
        acc = f(acc.clone(), xs[i].clone())
    }
```

so switching from the range version to the iterator version sounds good to me.